core/templates/services/blog-search.service.ts

```diff
@@ -0,0 +1,232 @@
+// Copyright 2022 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Search service for Blog Posts.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable, EventEmitter } from '@angular/core';
+
+import { BlogHomePageBackendApiService, SearchResponseData } from 'domain/blog/blog-homepage-backend-api.service';
+import cloneDeep from 'lodash/cloneDeep';
+
+export interface UrlSearchQuery {
+  searchQuery: string;
+  selectedTags: string[];
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class BlogPostSearchService {
+  // These properties are initialized using functions
+  // and we need to do non-null assertion. For more information, see
+  // https://github.com/oppia/oppia/wiki/Guide-on-defining-types#ts-7-1
+  private _lastQuery!: string;
+  private _searchOffset!: number | null;
+  private _lastSelectedTags: string[] = [];
+  private _isCurrentlyFetchingResults = false;
+  private _searchBarLoadedEventEmitter = new EventEmitter<string>();
+  private _initialSearchResultsLoadedEventEmitter =
+    new EventEmitter<SearchResponseData>();
+
+  public numSearchesInProgress = 0;
+
+  constructor(
+    private _blogHomePageBackendApiService: BlogHomePageBackendApiService,
+  ) {}
+
+  private _getSuffixForQuery(selectedTags: string[]): string {
+    let querySuffix = '';
+
+    let _tags = '';
+    for (let tag of selectedTags) {
+      if (_tags) {
+        _tags += '" OR "';
+      }
+      _tags += tag;
+    }
+    if (_tags) {
+      querySuffix += '&tags=("' + _tags + '")';
+    }
+    return querySuffix;
+  }
+
+  hasReachedLastPage(): boolean {
+    return this._searchOffset === null;
+  }
+
+  _updateSearchFields(itemsType: string, urlComponent: string): string[] {
```

Comment:
Please add docstring for this function to explain what it does.

Reply:
Changed function name and added doc string.